### ISSUE TITLE
Network outage causes Session to freeze

### ISSUE THREAD

**Issue by create-your-account** [lucas-clemente/quic-go#2258]
This seems to be caused by the internal send buffer of the UDP socket becoming full causing Linux kernel to return EAGAIN for writes. The golang netpoller handles this by waiting until the socket is writable ([source](https://github.com/golang/go/blob/release-branch.go1.13/src/internal/poll/fd_unix.go#L331)). This eventually causes Session to freeze on SendQueue.Send().

**Comment by marten-seemann** [comment]
This is an interesting issue, @create-your-account. The session should definitely not block in this case.

I just talked about this issue with @lucas-clemente, and we came up with the following idea: When sending a packet, we set a deadline of ~1ms. If the packet is not sent out within that time, this is a similar situation to congestion in the network (just that the congestion is closer to our side in this case). We then discard it, and report it as lost to the congestion controller, leading to a decrease of the congestion window and a thereby a slower send rate.

Does this make sense to you?

**Comment by create-your-account** [comment]
If I understand correctly in your proposal the SendQueue.Send() would block in case of outage for the duration of the deadline. I think such blocking should be avoided since when developing modifications to quic-go the developer might need to be aware of this special case of blocking.

To avoid having to ever wait on Send() I came up with another way to solve this. Checking in SendQueue.Send() if the internal queue is full. When queue is full report new packet as lost.

``` go
func (h *sendQueue) Send(p *packedPacket) {
	if len(h.queue) == cap(h.queue) {
		// report p as lost
                return
	}
	h.queue <- p
}
```
Downsides include the need to increase size of the queue to a value that would not cause packets to be dropped if for a short while packets are queued faster than consumed. Upside would be that inserting to the queue will never cause waiting.

One less portable solution for Linux would be before sending to use ioctl for SIOCOUTQ that returns number of data bytes in the local send queue and compare it to the value of SO_SNDBUF.

For my use case your proposal makes sense and has no reasonable issues.

**Comment by create-your-account** [comment]
For context mobile networks tend to fill the send buffer often during "normal" operation. One common cause is handover during change of base station. Also one ISP in my tests stops uplink for ~100ms in periods of 10 seconds. With other anomalies of mobile network included the send buffer of default size becomes full multiple times in one minute.

**Comment by create-your-account** [comment]
Maybe separate issue that causes deadlocks.

When Session is blocking on a full h.queue and sendQueue.Run() terminates with error, deadlock ensues. This is because the h.queue is never received from or closed now that Run() terminated. The h.queue is closed when Session is closing but for that to happen the Session needs to first receive the error from s.closeChan. This does not happen because Session is blocking on sendQueue.Send()

I wrote a small proof of concept that showcases the deadlock. By supplying a custom net.PacketConn to the client when dialing we can simulate WriteTo erroring after allowing h.queue to fill. The client's `Session.Close()` never returns due to the deadlock.

``` golang
type errPacketConn struct {
	wrapped  net.PacketConn
	writeErr chan error
}

func (c *errPacketConn) ReadFrom(p []byte) (n int, addr net.Addr, err error) {
	return c.wrapped.ReadFrom(p)
}

func (c *errPacketConn) WriteTo(p []byte, addr net.Addr) (n int, err error) {
	select {
	case err := <-c.writeErr:
		// Cause h.queue to fill
		time.Sleep(time.Second)
		return 0, err
	default:
	}
	return c.wrapped.WriteTo(p, addr)
}

func (c *errPacketConn) Close() error {
	return c.wrapped.Close()
}

func (c *errPacketConn) LocalAddr() net.Addr {
	return c.wrapped.LocalAddr()
}

func (c *errPacketConn) SetDeadline(t time.Time) error {
	panic("not implemented")
}

func (c *errPacketConn) SetReadDeadline(t time.Time) error {
	panic("not implemented")
}

func (c *errPacketConn) SetWriteDeadline(t time.Time) error {
	panic("not implemented")
}

func main() {
	ln, err := quic.ListenAddr(
		"127.0.0.1:",
		getTLSConfig(),
		&quic.Config{},
	)
	if err != nil {
		panic(err)
	}

	go func() {
		session, err := ln.Accept(context.Background())
		if err != nil {
			panic(err)
		}
		defer session.Close()

		stream, err := session.AcceptStream(context.Background())
		if err != nil {
			panic(err)
		}

		if _, err := io.Copy(ioutil.Discard, stream); err != nil {
			panic(err)
		}
	}()

	pconn, err := net.ListenPacket("udp", "127.0.0.1:")
	if err != nil {
		panic(err)
	}

	errConn := &errPacketConn{
		wrapped:  pconn,
		writeErr: make(chan error),
	}

	session, err := quic.Dial(errConn, ln.Addr(), ln.Addr().String(), getTLSClientConfig(), &quic.Config{})
	if err != nil {
		panic(err)
	}

	go func() {
		stream, err := session.OpenStream()
		if err != nil {
			panic(err)
		}

		// Make sure we cause writes
		ticker := time.NewTicker(100 * time.Millisecond)
		defer ticker.Stop()

		for range ticker.C {
			if _, err := stream.Write([]byte{0}); err != nil {
				panic(err)
			}
		}
	}()

	println("Causing deadlock...")
	errConn.writeErr <- errors.New("Deadlock")

	// Allow sendQueue to terminate
	time.Sleep(100 * time.Millisecond)

	println("Closing...")
	panic(session.Close())
}
```
